modules/miner/blockmanager_test.go

```diff
@@ -1,115 +1,227 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Get a header and solve it.
+	header, target, err := mt.miner.HeaderForWork()
+	if err != nil {
+		t.Fatal(err)
+	}
+	solvedHeader := solveHeader(header, target)
+	// Sanity check - header and solvedHeader should be different. (within the
+	// testing file, 'header' should always have a nonce of '0' and
+	// solvedHeader should never have a nonce of '0'.
+	if header.Nonce == solvedHeader.Nonce {
+		t.Fatal("nonce memory is not independent")
+	}
+
+	// Submit the header.
+	err = mt.miner.SubmitHeader(solvedHeader)
 	if err != nil {
-		return types.BlockHeader{}, err
+		t.Fatal(err)
 	}
-	*b, _ = m.SolveBlock(*b, m.target)
-	return b.Header(), nil
 }
 
-// TestBlockManager creates a miner, then polls the Miner for block
-// headers to mine. It ensures that the blockmanager properly hands
-// out headers, then reconstructs the blocks
-func TestBlockManager(t *testing.T) {
+// TestIntegrationHeaderForWorkUpdates checks that HeaderForWork starts
+// returning headers on the new block after a block has been submitted to the
+// consensus set.
+func TestIntegrationHeaderForWorkUpdates(t *testing.T) {
 	if testing.Short() {
 		t.SkipNow()
 	}
-	mt, err := createMinerTester("TestBlockManager")
+	mt, err := createMinerTester("TestIntegreationHeaderForWorkUpdates")
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	// Once we have polled for 2*headerForWorkMemory headers, the first
-	// headerForWorkMemory headers should be overwritten and no longer
-	// work.
-	headers := make([]types.BlockHeader, 2*headerForWorkMemory)
+	// Get a header to advance into the header memory.
+	_, _, err = mt.miner.HeaderForWork()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Submit a block, which should trigger a header change.
+	_, err = mt.miner.AddBlock()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Get a header to grind on.
+	header, target, err := mt.miner.HeaderForWork()
+	if err != nil {
+		t.Fatal(err)
+	}
+	solvedHeader := solveHeader(header, target)
+
+	// Submit the header.
+	err = mt.miner.SubmitHeader(solvedHeader)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !mt.cs.InCurrentPath(types.BlockID(crypto.HashObject(solvedHeader))) {
+		t.Error("header from solved block is not in the current path")
+	}
+}
+
+// TestIntegrationManyHeaders checks that requesting a thousand headers in
+// a row results in a thousand unique headers, and all of them can be
+// reassebled into valid blocks.
+func TestIntegrationManyHeaders(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	mt, err := createMinerTester("TestIntegrationManyHeaders")
+	if err != nil {
+		t.Fatal(err)
+	}
 
-	for i := 0; i < headerForWorkMemory; i++ {
-		headers[i], _, err = mt.miner.HeaderForWork()
+	// Create a suite of headers for imaginary parallel mining.
+	headerDepth := HeaderMemory / BlockMemory * 2
+	var solvedHeaders []types.BlockHeader
+	for i := 0; i < headerDepth; i++ {
+		header, target, err := mt.miner.HeaderForWork()
```

Comment:
it looks like you know the length of the slice in advance here, so you could write:
```go
solvedHeader := make([]types.BlockHeader, headerDepth)
for i := range solvedHeaders {
    header, target, err := mt.miner.HeaderForWork()
    if err != nil {
        t.Fatal(err)
    }
    solvedHeaders[i] = solveHeader(header, target)
}
```